R/aggregate.R

```diff
@@ -0,0 +1,166 @@
+    ## over all peptides instead of reference level.
+    X <- stats::model.matrix(~ -1 + sample + feature,
+                             contrasts.arg = list(feature = 'contr.sum'))
+    ## MASS::rlm breaks on singulare values.
+    ## - Check with base lm if singular values are present.
+    ## - If so, these coefficients will be zero, remove this collumn
+    ##   from model matrix
+    ## - Rinse and repeat on reduced modelmatrx till no singular
+    ##   values are present
+    repeat {
+        fit <- stats::.lm.fit(X, expression)
+        id <- fit$coefficients != 0
+        X <- X[ , id, drop = FALSE]
+        if (!any(!id)) break
+    }
+    ## Last step is always rlm: calculate estimated effects effects as
+    ## summarised values
+    fit <- MASS::rlm(X, expression, ...)
+
+    sampleid <- seq_along(unique(sample))
+    ## This will be needed for NUSE-type of quality control, but will
+    ## need to check for missing data as below.
+    ## se <- unique(summary(fit)$coefficients[sampleid, 'Std. Error'])
+
+    ## Take the sample coefficients ( = summarised expression values)
+    coef  <-  fit$coefficients[sampleid]
+    ## Sort the sample coefficients in the same way as the samplenames
+    ## of expression matrix. Puts NA for the samples without any
+    ## expression value
+    res <- coef[paste0('sample', colnames(x))]
+    names(res) <- colnames(x)
+    res
+}
+
+##' Fits an additive model (two way decomposition) using Tukey's median
+##' polish procedure using [stats::medpolish()].
+##'
+##' @title Return the Median Polish (Robust Twoway Decomposition) of a matrix
+##' @param x A numeric matrix.
+##' @param verbose Default is `FALSE`.
+##' @param ... Additional arguments passed to [stats::medpolish()].
+##' 
+##' @return A vector of numerics of length identical to `ncol(x)`.
+##' 
+##' @author Laurent Gatto
+##' 
+##' @export
+##'
+##' @importFrom stats medpolish
+##' 
+##' @examples
+##' x <- matrix(rnorm(30), nrow = 3)
+##' medianPolish(x)
+medianPolish <- function(x, verbose = FALSE, ...) {
+    medpol <- stats::medpolish(x, trace.iter = verbose, ...)
+    return(medpol$overall + medpol$col)
+}
+
+
+##' This function takes a matrix of quantitative features `x` and a
+##' factor (of length equal to `nrow(x)`) defining subsets, and
+##' applies a user-defined function to aggregate each subset into a
+##' vector of quantitative values.
+##'
+##' User-defined functions must thus return a vector of length equal
+##' to `ncol(x)`. Examples thereof are
+##'
+##' - [medianPolish()] to fits an additive model (two way decomposition)
+##'   using Tukey's median polish_ procedure using
+##'   [stats::medpolish()];
+##'
+##' - [robustSummary()] to calculate a robust aggregation using
+##'   [MASS::rlm()];
+##'
+##' - [base::colMeans()] to use the mean of each column;
+##'
+##' - [base::colSums()] to use the sum of each column;
+##'
+##' - [matrixStats::colMedians()] to use the median of each column.
```

Comment:
Is this working in `R CMD check` without a warning? Even if `matrixStats` is not in `Suggests`?

Reply:
Yes, at least on my computer. Will double check.